File: src/Wc3HandleLua/Handle/Image/Line.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local Color = Wc3Utils.Color or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')
local Log = Wc3Utils.Log or error('')

---@type ImagePixelClass
local Pixel = require('Handle.Image.Pixel') or error('')
---@type HandleSettings
local Settings = require('Settings') or error('')

--=======
-- Class
--=======

local ImageLine = Class.new('ImageLine')
---@class ImageLine
local public = ImageLine.public
---@class ImageLineClass
local static = ImageLine.static
---@type ImageLineClass
local override = ImageLine.override
local private = {}

--=========
-- Static
--=========

---@param max_p_count number
---@param p_size number | nil
---@param step number | nil
---@param child ImageLine | nil
---@return ImageLine
function override.new(max_p_count, p_size, step, child)
    isTypeErr(max_p_count, 'number', 'max_p_count')
    if p_size then isTypeErr(p_size, 'number', 'p_size') end
    if step then isTypeErr(step, 'number', 'step') end
    if child then isTypeErr(child, ImageLine, 'child') end

    local instance = child or Class.allocate(ImageLine)
    private.newData(instance, max_p_count, p_size, step)

    return instance
end

--========
-- Public
--========

---@param sx number
---@param sy number
---@param ex number
---@param ey number
function public:setPos(sx, sy, ex, ey)
    isTypeErr(self, ImageLine, 'self')
    isTypeErr(sx, 'number', 'sx')
    isTypeErr(sy, 'number', 'sy')
    isTypeErr(ex, 'number', 'ex')
    isTypeErr(ey, 'number', 'ey')

    local length = ((ex - sx)^2 + (ey - sy)^2)^0.5
    local angle = math.atan(ey - sy, ex - sx)
    self:setPosPolar(sx, sy, length, angle)
end

---@param x0 number
---@param y0 number
---@param length number
---@param angle number
function public:setPosPolar(x0, y0, length, angle)
    isTypeErr(self, ImageLine, 'self')
    isTypeErr(x0, 'number', 'x0')
    isTypeErr(y0, 'number', 'y0')
    isTypeErr(length, 'number', 'length')
    isTypeErr(angle, 'number', 'angle')
    local priv = private.data[self]

    if not priv.is_visible then
        return
    end

    local need = math.floor(length / priv.step) + 1
    if need > #priv.pixels then
        Log:wrn('ImageLine: not enought pixels to fill.')
    end

    local step_x = priv.step * math.cos(angle)
    local step_y = priv.step * math.sin(angle)

    local used = math.min(#priv.pixels, need)
    for i = 1, used do
        local pixel = priv.pixels[i]
        pixel:setPos(x0, y0)
        pixel:setVisible(true)

        x0 = x0 + step_x
        y0 = y0 + step_y
    end

    for i = used, #priv.pixels do
        priv.pixels[i]:setVisible(false)
    end
end

---@param color Color
function public:setColor(color)
    isTypeErr(self, ImageLine, 'self')
    isTypeErr(color, Color, 'color')
    local priv = private.data[self]

    priv.color = Color.copy(color)
    for i = 1, #priv.pixels do
        priv.pixels[i]:setColor(color)
    end
end

---@return Color
function public:getColor()
    isTypeErr(self, ImageLine, 'self')
    return Color.copy(private.data[self].color)
end

---@param flag boolean
function public:renderAlways(flag)
    isTypeErr(self, ImageLine, 'self')
    isTypeErr(flag, 'boolean', 'flag')
    local priv = private.data[self]

    priv.render_always = flag
    for i = 1, #priv.pixels do
        priv.pixels[i]:renderAlways(flag)
    end
end

---@param flag boolean
function public:setVisible(flag)
    isTypeErr(self, ImageLine, 'self')
    isTypeErr(flag, 'boolean', 'flag')
    local priv = private.data[self]

    priv.is_visible = flag
    for i = 1, #priv.pixels do
        priv.pixels[i]:setVisible(flag)
    end
end

function public:destroy()
    isTypeErr(self, ImageLine, 'self')
    local priv = private.data[self]

    for i = 1, #priv.pixels do
        priv.pixels[i]:destroy()
    end
end

--=========
-- Private
--=========

private.data = setmetatable({}, {__mode = 'k'})

---@param self ImageLine
---@param max_p_count number
---@param p_size number
---@param step number
function private.newData(self, max_p_count, p_size, step)
    local priv = {
        color = Color.new(1, 1, 1, 1),
        render_always = true,
        is_visible = false,

        step = step or Settings.ImagePixelDefaultSize / 2,
        pixels = {}
    }
    private.data[self] = priv

    for i = 1, max_p_count do
        local pixel = Pixel.new(p_size)
        pixel:setColor(priv.color)
        pixel:renderAlways(priv.render_always)
        pixel:setVisible(priv.is_visible)

        table.insert(priv.pixels, pixel)
    end
end

return static

File: src/Handle/Image/Line.ts
import { Color, Log} from '../../Utils'
import { hPixel, hPixelList } from "./Pixel";

export class hImageLine<T extends hPixel> {
    constructor(pixels: hPixelList<T>,
                pixel_step?: number){

        this._step = pixel_step ? pixel_step : 8
        this._in_use = 0
        this._pixels = pixels
    }

    setPos(x1: number, y1: number, x2: number, y2: number){
        this._x1 = x1
        this._y1 = y1
        this._x2 = x2
        this._y2 = y2

        let dx = x2 - x1
        let dy = y2 - y1
        this._length = SquareRoot(dx * dx + dy * dy)
        this._angle = Atan2(dy, dx)

        this.update()
    }

    setPolarPos(x1: number, y1: number, length: number, angle: number){
        this._x1 = x1
        this._y1 = y1
        this._x2 = x1 + length * Cos(angle)
        this._y2 = y1 + length * Sin(angle)
        this._length = length
        this._angle = angle

        this.update()
    }

    setPolarPos2(cx: number, cy: number, r1: number, a1: number, r2: number, a2: number){
        this._x1 = cx + r1 * Cos(a1)
        this._y1 = cy + r1 * Sin(a1)
        this._x2 = cx + r2 * Cos(a2)
        this._y2 = cy + r2 * Sin(a2)

        let dx = this._x2 - this._x1
        let dy = this._y2 - this._y1
        this._length = SquareRoot(dx * dx + dy * dy)
        this._angle = Atan2(dy, dx)

        this.update()
    }

    get x1(){return this._x1}
    get y1(){return this._y1}
    get x2(){return this._x2}
    get y2(){return this._y2}
    get length(){return this._length}
    get angle(){return this._angle}
    
    get color(){return new Color(this._color)}
    set color(color: Color){
        this._color = new Color(color)
        for (let pixel of this._pixels){
            pixel.color = color
        }
    }

    get visible(){return this._visible}
    set visible(flag: boolean){
        this._visible = flag
        for (let i = 0; i < math.min(this._in_use, this._pixels.length); i++){
            this._pixels[i].visible = flag
        }
    }

    destroy(){
        for (let pixel of this._pixels){
            pixel.destroy()
        }
    }

    private update(){
        this._in_use = math.floor(this._length / this._step) + 1
        if (this._in_use > this._pixels.length){
            Log.wrn(hImageLine.name + 
                    ': is not enought pixels to fill line.')
        }

        let x = this._x1
        let y = this._y1
        let step_x = this._step * Cos(this._angle)
        let step_y = this._step * Sin(this._angle)
        let rly_used = math.min(this._in_use, this._pixels.length)

        for (let i = 0; i < rly_used; i++){
            let pixel = this._pixels[i]
            pixel.visible = this._visible
            pixel.x = x
            pixel.y = y
            x += step_x
            y += step_y
        }

        for (let i = rly_used; i < this._pixels.length; i++){
            this._pixels[i].visible = false
        }
    }

    private _x1: number = 0
    private _y1: number = 0
    private _x2: number = 0
    private _y2: number = 0
    private _length: number = 0
    private _angle: number = 0

    private _step: number
    private _in_use: number
    private _pixels: hPixelList<T>

    private _color = new Color(1, 1, 1, 1)
    private _render_always = true
    private _visible = false
}